**Special Tokens in Tokenizers**

Natural language tokenizers use special tokens to handle certain tasks such as padding, classification, and separating segments. Below are their meanings:

[PAD]: Used to make all input sequences the same length by adding padding at the end.

[UNK]: Represents unknown words not found in the vocabulary.

[CLS]: Placed at the beginning of every input for classification tasks.

[SEP]: Separates multiple sentences or segments in one input.

[MASK]: Used in training to hide a word so the model learns to predict it.

[unusedX]: Reserved tokens not currently used but available for custom use.



---

**What is BERT ?**

🧒👕 .. Sesame Street :)


BERT, short for Bidirectional Encoder Representations from Transformers, is a machine learning (ML) framework for natural language processing


---



In [1]:
from transformers import BertTokenizer
# The code loads the BERT tokenizer and analyzes it
tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")

def analyze_tokenizer(tokenizer):
    # Print the total vocabulary size in the form (the number of words it knows)
    vocab_size = tokenizer.vocab_size
    print(f"Vocabulary size: {vocab_size}")
    # Show the first 10 symbols in the vocabulary, just for a look 👀
    sample_vocab = list(tokenizer.vocab.keys())[:10]
    print(f"Sample of the vocabulary: {sample_vocab}")
    # Print special characters and their IDs within vocabulary
    print("Special tokens:")
    print(f" - [PAD] token: {tokenizer.pad_token}, ID: {tokenizer.pad_token_id}")
    print(f" - [CLS] token: {tokenizer.cls_token}, ID: {tokenizer.cls_token_id}")
    print(f" - [SEP] token: {tokenizer.sep_token}, ID: {tokenizer.sep_token_id}")
    print(f" - [MASK] token: {tokenizer.mask_token}, ID: {tokenizer.mask_token_id}")
    # A text example that we pass to the encoder to convert it to encoded tokens (input_ids)
    sample_text = "Pizza is my  weakness. No regrets"
    encoded = tokenizer(sample_text)
    # Print Original Text
    # Print Tokenization
    # Print Input_ids
    print(f"\nSample text: {sample_text}")
    print(f"Tokenized text: {tokenizer.tokenize(sample_text)}")
    print(f"Encoded IDs: {encoded['input_ids']}")
    # Convert the digital identifiers back to text to see how the model understands the text
    decoded_text = tokenizer.decode(encoded['input_ids'])
    print(f"Decoded text: {decoded_text}")

# Run the analysis
analyze_tokenizer(tokenizer)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

Vocabulary size: 30522
Sample of the vocabulary: ['[PAD]', '[unused0]', '[unused1]', '[unused2]', '[unused3]', '[unused4]', '[unused5]', '[unused6]', '[unused7]', '[unused8]']
Special tokens:
 - [PAD] token: [PAD], ID: 0
 - [CLS] token: [CLS], ID: 101
 - [SEP] token: [SEP], ID: 102
 - [MASK] token: [MASK], ID: 103

Sample text: Pizza is my  weakness. No regrets
Tokenized text: ['pizza', 'is', 'my', 'weakness', '.', 'no', 'regrets']
Encoded IDs: [101, 10733, 2003, 2026, 11251, 1012, 2053, 23161, 102]
Decoded text: [CLS] pizza is my weakness. no regrets [SEP]


**Result after encryption (Encoded IDs):**

Encoded IDs: [101, 10733, 2003, 2026, 11251, 1012, 2053, 23161, 102]

This is the text after converting it to numbers (IDs), Where:

101 = [CLS]

102 = [SEP]

The others are the IDs for the words

 **Result after decryption (Decoded text):**

 Decoded text: [CLS] pizza is my weakness. no regrets [SEP]

 => The model was able to successfully convert numbers to words again, automatically adding the symbols [CLS] and [SEP]



**Result after encryption (Encoded IDs):**

Encoded IDs: [101, 10733, 2003, 2026, 11251, 1012, 2053, 23161, 102]

This is the text after converting it to numbers (IDs), Where:

101 = [CLS]

102 = [SEP]

The others are the IDs for the words

 **Result after decryption (Decoded text):**

 Decoded text: [CLS] pizza is my weakness. no regrets [SEP]

 => The model was able to successfully convert numbers to words again, automatically adding the symbols [CLS] and [SEP]




---

here we can see the **Case Normalization**

the BERT tokenizer uses a model called "bert-base-uncased", which performs case normalization. This means that all capital letters (uppercase) in the input text are converted to lowercase

"Pizza" to "pizza" 🍕  and "No" to "no" 🙅🏻‍♂️



In [2]:
import json

with open("Tokenizer.ipynb", "r", encoding="utf-8") as f:
    notebook_data = json.load(f)

print(notebook_data.get("metadata", "لا توجد بيانات وصفية"))

{'colab': {'provenance': []}, 'kernelspec': {'display_name': 'Python 3', 'name': 'python3'}, 'language_info': {'codemirror_mode': {'name': 'ipython', 'version': 3}, 'file_extension': '.py', 'mimetype': 'text/x-python', 'name': 'python', 'nbconvert_exporter': 'python', 'pygments_lexer': 'ipython3', 'version': '3.13.0'}, 'widgets': {'application/vnd.jupyter.widget-state+json': {'00bd9b71c48f467692957f0b92bce4fc': {'model_module': '@jupyter-widgets/base', 'model_module_version': '1.2.0', 'model_name': 'LayoutModel', 'state': {'_model_module': '@jupyter-widgets/base', '_model_module_version': '1.2.0', '_model_name': 'LayoutModel', '_view_count': None, '_view_module': '@jupyter-widgets/base', '_view_module_version': '1.2.0', '_view_name': 'LayoutView', 'align_content': None, 'align_items': None, 'align_self': None, 'border': None, 'bottom': None, 'display': None, 'flex': None, 'flex_flow': None, 'grid_area': None, 'grid_auto_columns': None, 'grid_auto_flow': None, 'grid_auto_rows': None, 'g

In [4]:
with open("C:/Users/suras/Downloads/Tokenizer.ipynb", "r", encoding="utf-8") as f:
    notebook_data = json.load(f)

In [6]:
notebook_data["metadata"]["widgets"] = {"state": {}}

with open("Tokenizer.pynb", "w", encoding="utf-8") as f:
    json.dump(notebook_data, f, indent=4)